WordPress/gutenberg
# Make the Preview Menu extensible

## Description

This PR introduces a new `<PluginPreview>` component to the `@wordpress/block-editor` package. It allows for defining a custom "Preview" menu item along with optional content that is rendered instead of the `VisualEditor`.

Thanks to that, plugin authors can extend the "Preview" menu and provide optional custom content previews.

This PR is based on #25430 proposed by @mreishus. It addresses @jorgefilipecosta's [feedback](https://github.com/WordPress/gutenberg/pull/25430#discussion_r503445235) pointing out that the new component added to `@wordpress/block-editor` should not depend on the `@wordpress/edit-post` package.

In the original implementation we were using a `deviceType` and a `setDeviceType` action from the `core/edit-post` store. Right now, they are [passed](https://github.com/WordPress/gutenberg/compare/trunk...delawski:add/preview-menu-slot#diff-61a1980577fb29dd7a7ed4d0a4c23919d5c6bbae93517a9957f80e1c04d50392R106-R107) to the slot fill by the [`VisualEditorOrPluginPreview`](https://github.com/WordPress/gutenberg/compare/trunk...delawski:add/preview-menu-slot#diff-4ea0366d68b391dbc2c4825aeba64fc12057883203c03bfe0a50c9ebecf4e51bR28) component (that lives in the `@wordpress/edit-post` package).

There's one additional difference between this PR and the [original one](#25430). The custom preview (`PluginPreview` component children) is optional now. If there is no custom preview provided, selecting a custom "Preview" menu item will not affect the `VisualEditor`. Instead, plugin authors may rely on the `onClick` event handler for providing custom functionality. This may be very helpful in the case of plugins that provide external previews (e.g. paywalled content, search results, or AMP preview as proposed in ampproject/amp-wp#5943).

## How has this been tested?



- Create a new block editor plugin as follows:

```js
import domReady from '@wordpress/dom-ready';
import { __ } from '@wordpress/i18n';
import { PluginPreview } from '@wordpress/block-editor';
import { registerPlugin } from '@wordpress/plugins';
import { external } from '@wordpress/icons';

domReady( () => {
	const PluginPreviewTest = () => (
		<>
			<PluginPreview
				name="preview-custom-1"
				title={ __( 'Custom Preview 1' ) }
			>
				<h1>
					{ __( 'Custom Preview 1 Content' ) }
				</h1>
			</PluginPreview>

			<PluginPreview
				name="preview-custom-2"
				title={ __( 'Custom Preview 2' ) }
				onClick={ ( event ) => console.log( event  ) }
			>
				<h1>
					{ __( 'Custom Preview 2 Content' ) }
				</h1>
			</PluginPreview>

			<PluginPreview
				name="custom-action"
				title={ __( 'Custom Action' ) }
				icon={ external }
				onClick={ ( event ) => console.log( event ) }
			/>
		</>
	);

	registerPlugin( "plugin-preview-test", {
		render: PluginPreviewTest,
	} );
} );
```
- Go to block editor and confirm there are 3 new items in the "Preview" menu:
![Screenshot 2021-05-19 at 11 38 23](https://user-images.githubusercontent.com/478735/118791168-bff45c80-b896-11eb-80f5-55e7e73ccc90.png)
- Click the "Custom Preview 1" menu item and confirm that the block editor content has been replaced with a custom preview
- Click the "Custom Preview 2" menu item and confirm that besides changing the editor content, a click event object has been logged to the console
![Screenshot 2021-05-19 at 11 40 19](https://user-images.githubusercontent.com/478735/118791450-034ecb00-b897-11eb-93bf-f351071df650.png)
- Click the "Custom Action" menu item and confirm the editor config has not been changed while a click event has been logged to the console.

## Screenshots 

![Screenshot 2021-05-18 at 20 50 27](https://user-images.githubusercontent.com/478735/118790681-4bb9b900-b896-11eb-8828-035af3a5f0c2.png)

## Types of changes
New feature (non-breaking change which adds functionality)

Fixes #25309
Relates to Automattic/wp-calypso#42896
Relates to ampproject/amp-wp#5943

## Checklist:
